embedded-graphics/src/primitives/line.rs

```diff
@@ -90,121 +90,84 @@ where
-
-        // Find out if our line is steep or shallow
-        let is_steep = (y1 - y0).abs() > (x1 - x0).abs();
-
-        // Determine if endpoints should be switched
-        // based on the "quick" direction
-        let (x0, y0, x1, y1) = if is_steep {
-            if y0 > y1 {
-                (x1, y1, x0, y0)
-            } else {
-                (x0, y0, x1, y1)
-            }
-        } else {
-            if x0 > x1 {
-                (x1, y1, x0, y0)
-            } else {
-                (x0, y0, x1, y1)
-            }
-        };
-
-        // Setup our pre-calculated values
-        let (dquick, mut dslow) = if is_steep {
-            (y1 - y0, x1 - x0)
-        } else {
-            (x1 - x0, y1 - y0)
-        };
+        let mut d = self.end - self.start;
+        if d[0] < 0 {
+            d = Coord::new(-d[0], d[1]);
+        }
+        if d[1] > 0 {
+            d = Coord::new(d[0], -d[1]);
+        }
 
-        // Determine how we should increment the slow direction
-        let increment = if dslow < 0 {
-            dslow = -dslow;
-            -1
-        } else {
-            1
+        let s = match (self.start[0] >= self.end[0], 
+                       self.start[1] >= self.end[1]) {
+            (false, false) => Coord::new(1, 1),
+            (false, true) => Coord::new(1, -1),
+            (true, false) => Coord::new(-1, 1),
+            (true, true) => Coord::new(-1, -1),
         };
 
-        // Compute the default error
-        let error = 2 * dslow - dquick;
-
-        // Set our inital quick & slow
-        let (quick, slow, end) = if is_steep { (y0, x0, y1) } else { (x0, y0, x1) };
-
         LineIterator {
-            line: self,
-
-            is_steep,
-            dquick,
-            dslow,
-            increment,
-            error,
-
-            quick,
-            slow,
-            end,
+            style: self.style,
+            
+            start: self.start,
+            end: self.end,
+            d,
+            s,
+            err: d[0] + d[1],
+            e2: 0,
+            stop: false,
         }
     }
 }
 
 /// Pixel iterator for each pixel in the line
-#[derive(Debug)]
-pub struct LineIterator<'a, C: 'a>
+#[derive(Debug, Clone, Copy)]
+pub struct LineIterator<C>
 where
     C: PixelColor,
 {
-    line: &'a Line<C>,
-
-    dquick: i32,
-    dslow: i32,
-    increment: i32,
-    error: i32,
-    is_steep: bool,
-
-    quick: i32,
-    slow: i32,
-    end: i32,
+    style: Style<C>,
+
+    start: Coord,
+    end: Coord,
+    d: Coord,
+    s: Coord,
+    err: i32,
+    e2: i32,
+    stop: bool,
 }
 
 // [Bresenham's line algorithm](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm)
-impl<'a, C: PixelColor> Iterator for LineIterator<'a, C> {
+impl<C: PixelColor> Iterator for LineIterator<C> {
     type Item = Pixel<C>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.quick > self.end {
-            return None;
-        }
+        self.style.stroke_color?;
```

Comment:
Can you add a quick comment explaining that this bails early if there's no stroke? It took me a moment to figure out so might confuse others too. Quite an elegant solution though, nice!